### PR TITLE
Add WriterUuid support

### LINKED ISSUE #177: Implement WriterUuid feature

### PR DESCRIPTION
**Pull request by Arkatufus**
Fixes #177

## Changes

- Add a randomly generated uuid (using GUID) to write journal
- Add a "writer_uuid" column to journal table
- Make sure that writer uuid is written to the table
- Add unit test to make sure that default tables and column names are generated properly

### CODE REVIEW

**Comment by Arkatufus** [comment]
NOTE: we currently do not have a proper unit test to make sure that writer uuid are written to the journal table

File: src/Akka.Persistence.Sql/Config/EventJournalTableConfig.cs [hunk 1]
```diff
@@ -10,13 +10,16 @@ namespace Akka.Persistence.Sql.Config
 {
     public sealed class EventJournalTableConfig : IEquatable<EventJournalTableConfig>
     {
-        public EventJournalTableConfig(Configuration.Config config)
+        public EventJournalTableConfig(Configuration.Config config, string tableMapName)
```
Comment (Arkatufus) [inline src/Akka.Persistence.Sql/Config/EventJournalTableConfig.cs]: We need to know if we're using the default table mapping or if we're running in a backward compat mode. WriterUuid is only available in Akka.Persistence.Sql and not in any other legacy plugins.

File: src/Akka.Persistence.Sql/Config/JournalTableColumnNames.cs [hunk 2]
```diff
@@ -22,6 +22,7 @@ public JournalTableColumnNames(Configuration.Config config)
             Message = cfg.GetString("message", "message");
             Identifier = cfg.GetString("identifier", "identifier");
             Manifest = cfg.GetString("manifest", "manifest");
+            WriterUuid = cfg.GetString("writer-uuid", "writer_uuid");
```
Comment (Arkatufus) [inline src/Akka.Persistence.Sql/Config/JournalTableColumnNames.cs:25]: New column HOCON config property `WriterUuid`

File: src/Akka.Persistence.Sql/Db/AkkaPersistenceDataConnectionFactory.cs [hunk 3]
```diff
@@ -136,13 +136,24 @@ public AkkaPersistenceDataConnectionFactory(IProviderConfig<SnapshotTableConfigu
...
                 .HasColumnName(columnNames.Created)
 
-                // TODO: Disabling this for now, will need a migration script to support this
-                .Member(r => r.WriteUuid)
-                .IsNotColumn()
-
                 .Member(r => r.TagArr)
                 .IsNotColumn();
 
+            if (journalConfig.TableMapName == "default")
```
Comment (Arkatufus) [inline src/Akka.Persistence.Sql/Db/AkkaPersistenceDataConnectionFactory.cs]: We only create the `WriterUuid` column if we're creating a new default database (`Akka.Persistence.Sql` only), and we do not offer any backward compatibility check.

File: src/Akka.Persistence.Sql/Journal/Dao/BaseByteArrayJournalDao.cs [hunk 4]
```diff
@@ -29,38 +29,6 @@
 
 namespace Akka.Persistence.Sql.Journal.Dao
 {
-    public static class SequentialUuidGenerator
```
Comment (Arkatufus) [inline src/Akka.Persistence.Sql/Journal/Dao/BaseByteArrayJournalDao.cs:33]: Remove the old Uuid generator, we'll be using the default randomized Guid instead.

File: src/Akka.Persistence.Sql/Journal/SqlWriteJournal.cs [hunk 5]
```diff
@@ -54,6 +54,7 @@ public class SqlWriteJournal : AsyncWriteJournal
         public SqlWriteJournal(Configuration.Config journalConfig)
         {
             _log = Context.GetLogger();
+            var selfUuid = Guid.NewGuid().ToString("N");
```
Comment (Arkatufus) [inline src/Akka.Persistence.Sql/Journal/SqlWriteJournal.cs]: Actual implementation, writer uuid is generated every time the write journal is instantiated.

File: src/Akka.Persistence.Sql/persistence.conf [hunk 6]
```diff
@@ -121,6 +121,7 @@
             message = message
             identifier = identifier
             manifest = manifest
+            writer-uuid = writer_uuid
```
Comment (Arkatufus) [inline src/Akka.Persistence.Sql/persistence.conf:147]: New column mapping for writer uuid

File: src/Akka.Persistence.Sql/Journal/Dao/ByteArrayJournalSerializer.cs [hunk 7]
```diff
@@ -58,15 +61,17 @@ private static string StringSep(IImmutableSet<string> tags, string separator)
...
             => tagWriteMode switch
             {
                 TagMode.Csv => new JournalRow
                 {
                     Tags = StringSep(tags, separator),
                     Timestamp = representation.Timestamp == 0
                         ? timestamp
-                        : representation.Timestamp
+                        : representation.Timestamp,
+                    WriteUuid = uuid
```
Comment (Arkatufus) [inline src/Akka.Persistence.Sql/Journal/Dao/ByteArrayJournalSerializer.cs]: Actual implementation, WriterUuid property is now being set

File: src/Akka.Persistence.Sql/Journal/Dao/ByteArrayJournalSerializer.cs [hunk 8]
```diff
@@ -75,7 +80,8 @@ private static string StringSep(IImmutableSet<string> tags, string separator)
                     TagArr = tags.ToArray(),
                     Timestamp = representation.Timestamp == 0
                         ? timestamp
-                        : representation.Timestamp
+                        : representation.Timestamp,
+                    WriteUuid = uuid
```
Comment (Arkatufus) [inline src/Akka.Persistence.Sql/Journal/Dao/ByteArrayJournalSerializer.cs]: Actual implementation, WriterUuid property is now being set

File: src/Akka.Persistence.Sql/Journal/Dao/ByteArrayJournalSerializer.cs [hunk 9]
```diff
@@ -84,7 +90,8 @@ private static string StringSep(IImmutableSet<string> tags, string separator)
...
```
Comment (Arkatufus) [inline src/Akka.Persistence.Sql/Journal/Dao/ByteArrayJournalSerializer.cs]: Actual implementation, WriterUuid property is now being set

File: src/Akka.Persistence.Sql/Journal/Dao/ByteArrayJournalSerializer.cs [hunk 10]
```diff
@@ -150,7 +157,7 @@ protected override Try<(IPersistentRepresentation, IImmutableSet<string>, long)>
                             manifest: t.EventManifest ?? t.Manifest,
                             isDeleted: t.Deleted,
                             sender: ActorRefs.NoSender,
-                            writerGuid: null,
+                            writerGuid: t.WriteUuid,
```
Comment (Arkatufus) [inline src/Akka.Persistence.Sql/Journal/Dao/ByteArrayJournalSerializer.cs]: Actual implementation, WriterUuid column is now being persisted to database

File: src/Akka.Persistence.Sql/Journal/Dao/ByteArrayJournalSerializer.cs [hunk 11]
```diff
@@ -172,7 +179,7 @@ protected override Try<(IPersistentRepresentation, IImmutableSet<string>, long)>
                         manifest: t.EventManifest ?? t.Manifest,
                         isDeleted: t.Deleted,
                         sender: ActorRefs.NoSender,
-                        writerGuid: null,
+                        writerGuid: t.WriteUuid,
```
Comment (Arkatufus) [inline src/Akka.Persistence.Sql/Journal/Dao/ByteArrayJournalSerializer.cs]: Actual implementation, WriterUuid column is now being persisted to database

**Comment by Arkatufus** [comment]
Question: Should this feature be gated using HOCON configuration instead? let the user decide if they want to use WriterUuid and use true as default?

**Comment by Aaronontheweb** [comment]
> Question: Should this feature be gated using HOCON configuration instead? let the user decide if they want to use WriterUuid and use true as default?

definitely should be on by default - should only be off if the user can't / won't run the schema migration.

**Comment by Aaronontheweb** [comment]
> NOTE: we currently do not have a proper unit test to make sure that writer uuid are written to the journal table

We don't have one of these inside the TCK?

**Comment by Arkatufus** [comment]
> We don't have one of these inside the TCK?

No we do not, this "feature" has been neglected and has never been implemented in all of the SQL plugins

**Comment by Aaronontheweb** [comment]
> > We don't have one of these inside the TCK?
> 
> No we do not, this "feature" has been neglected and has never been implemented in all of the SQL plugins

Can you add an issue to add a spec for this to the TCK? And for the plugins that don't properly support it, make it possible for them to disable that spec. This really should be part of the core `JournalSpecs`.

**Comment by Aaronontheweb** [comment]
@Arkatufus you'll need to rebase this

**Comment by Arkatufus** [comment]
> Can you add an issue to add a spec for this to the TCK?

Will do

File: src/Akka.Persistence.Sql/Config/EventJournalTableConfig.cs [hunk 12]
```diff
@@ -14,11 +14,14 @@ public EventJournalTableConfig(Configuration.Config config)
         {
             var journalConfig = config.GetConfig("journal");
             Name = journalConfig.GetString("table-name");
+            UseWriterUuidColumn = journalConfig.GetBoolean("use-writer-uuid-column");
             ColumnNames = new JournalTableColumnNames(journalConfig);
         }
 
         public string Name { get; }
 
+        public bool UseWriterUuidColumn { get; }
```
Comment (Aaronontheweb) [inline src/Akka.Persistence.Sql/Config/EventJournalTableConfig.cs:23]: LGTM

File: src/Akka.Persistence.Sql/Config/JournalTableConfig.cs [hunk 13]
```diff
@@ -40,7 +40,7 @@ public JournalTableConfig(Configuration.Config config)
 
         public string SchemaName { get; }
 
-        // TODO: implement this settings
+        [Obsolete("Not implemented")]
         public bool UseEventManifestColumn { get; } = false;
```
Comment (Aaronontheweb) [inline src/Akka.Persistence.Sql/Config/JournalTableConfig.cs]: Why is this obsolete?

File: src/Akka.Persistence.Sql/Db/AkkaPersistenceDataConnectionFactory.cs [hunk 14]
```diff
@@ -136,13 +136,24 @@ public AkkaPersistenceDataConnectionFactory(IProviderConfig<SnapshotTableConfigu
...
                 .HasColumnName(columnNames.Created)
 
-                // TODO: Disabling this for now, will need a migration script to support this
-                .Member(r => r.WriteUuid)
-                .IsNotColumn()
-
                 .Member(r => r.TagArr)
                 .IsNotColumn();
 
+            if (journalConfig.UseWriterUuidColumn)
```
Comment (Aaronontheweb) [inline src/Akka.Persistence.Sql/Db/AkkaPersistenceDataConnectionFactory.cs:142]: LGTM

File: src/Akka.Persistence.Sql/Journal/Types/JournalRow.cs [hunk 15]
```diff
@@ -30,7 +30,7 @@ public sealed class JournalRow
 
         public string[] TagArr { get; set; }
 
-        public Guid? WriteUuid { get; set; }
+        public string WriterUuid { get; set; }
```
Comment (Aaronontheweb) [inline src/Akka.Persistence.Sql/Journal/Types/JournalRow.cs:33]: LGTM - probably easier to work with strings than Guids. We don't gain much from a Db perspective using a strict type here.

**Review by Aaronontheweb — COMMENTED** [review]
Overall looks good, just had a question for you

File: src/Akka.Persistence.Sql/Config/JournalTableConfig.cs (hunk 13 above)
Comment (Arkatufus) [reply src/Akka.Persistence.Sql/Config/JournalTableConfig.cs]: It was never implemented, configuration wise.

Comment (Arkatufus) [reply src/Akka.Persistence.Sql/Config/JournalTableConfig.cs]: It was me using Obsolete as a debugging tool, i'll turn it back on

**Pull request merged**